File: lambdify-apigateway-java/tests/lambdify/apigateway/MyApp.java
package lambdify.apigateway;

import static java.util.Collections.*;
import static lambdify.apigateway.Methods.*;
import lambdify.aws.events.apigateway.*;
import lombok.*;

public class MyApp extends App {{

    val users = new UserRepository();

    routes(
        GET.and( "/users" ).with( users::retrieveUsers ),
        GET.and("/users/:id" ).with( users::retrieveSingleUser ),
        PATCH.and("/users" ).with( users::createReportOfUsers ),
        POST.and("/users" ).withNoContent( users::saveUser )
    );
}}

class UserRepository {

    ProxyResponseEvent customNotFoundHandler(ProxyRequestEvent request ) {
        return Responses.notFound()
            .withHeaders( singletonMap( "X-Custom", "Not Found" ) );
    }

    ProxyResponseEvent retrieveUsers( ProxyRequestEvent request ) {
        return Responses.ok( singletonList(new User("User") ), "application/json" );
    }

    ProxyResponseEvent retrieveSingleUser( ProxyRequestEvent request ) {
        return Responses.ok( "{'name':'Lambda User'}", "application/json" );
    }

    ProxyResponseEvent createReportOfUsers(){
        return Responses.created();
    }

    void saveUser( ProxyRequestEvent request ) {}

    ProxyResponseEvent updateUser( ProxyRequestEvent request ){
        val user = RequestParameterReader.getBodyAs( request, User.class );
        return Responses.ok( user.getName(), "plain/text" );
    }

    @NoArgsConstructor
    @RequiredArgsConstructor
    @Getter @Setter public static class User {
        @NonNull String name;
    }
}

File: lambdify-apigateway-kotlin/tests/lambdify/apigateway/MyApp.kt
package lambdify.apigateway

import lambdify.apigateway.Methods.*
import lambdify.apigateway.kotlin.App
import lambdify.apigateway.kotlin.and
import lambdify.apigateway.kotlin.with
import lambdify.apigateway.kotlin.withNoContent
import lambdify.aws.events.apigateway.ProxyRequestEvent
import lambdify.aws.events.apigateway.ProxyResponseEvent

/**
 * Created by miere.teixeira on 06/04/2018.
 */
class MyApp: App({

    val users = UserResource()

    ApiGatewayConfig.INSTANCE.defaultNotFoundHandler( users::customNotFoundHandler )

    routes(
        GET and "/users" with users::retrieveUsers,
        GET and "/users/:id" with users::retrieveSingleUser,
        PATCH and "/users" with users::doReportOfUsers,
        POST and "/users" withNoContent users::saveUser
    )
})

open class UserResource {

    lateinit var lastExecutedMethod:String

    fun customNotFoundHandler( request: ProxyRequestEvent): ProxyResponseEvent {
        lastExecutedMethod = "customNotFoundHandler"
        return Responses.notFound().apply {
            this.headers = mutableMapOf(
                "X-Custom" to "Not Found"
            )
        }
    }

    fun retrieveUsers( request: ProxyRequestEvent): ProxyResponseEvent {
        lastExecutedMethod = "retrieveUsers"
        return Responses.noContent()
    }

    fun retrieveSingleUser( request: ProxyRequestEvent? ): ProxyResponseEvent {
        lastExecutedMethod = "retrieveSingleUser"
        return Responses.ok("{'name':'Lambda User'}", "application/json")
    }

    fun doReportOfUsers(): ProxyResponseEvent {
        lastExecutedMethod = "doReportOfUsers"
        return Responses.created()
    }

    fun saveUser( request: ProxyRequestEvent? ) {
        lastExecutedMethod = "saveUser"
    }
}